### DISCUSSION TITLE
more convenient property binding?

### DISCUSSION THREAD

**Discussion by mberndt123** [scalafx/scalafx#375] [Ideas]
Hi, 

I was looking at the `Bindings.createObjectBinding`.

One needs to specify the dependencies separately from the function that computes the value of the property. This seems a bit error-prone to me. One can easily specify dependencies that aren't needed l, leading to values being recomputed unnecessarily, or one can fail to specify dependencies that are used, meaning that a property would not be updated when that is necessary. I've done some amount of programming with Vue.js where that is taken care of automatically and I find it very helpful.

What do you think about adding some APIs to simplify this? An obvious one would be a mapN method that allows one to map the values of a number of inputs to an output.

Then again, maybe it’s not necessary to reinvent the wheel? I’m aware of at least 3 Scala libraries that implement this sort of functionality:
[Binding.scala](https://github.com/ThoughtWorksInc/Binding.scala), [RxScala](https://reactivex.io/rxscala/) and [scala.rx](https://github.com/lihaoyi/scala.rx). I think some sort of integration with one of these libraries would be very useful for ScalaFX!

**Comment by jpsacha** [comment]
Can you provide an example how those libraries can be used to improve `createObjectBinding` functionality?

**Comment by mberndt123** [comment]
Hey @jpsacha,

having thought more about it, I think that for ScalaFX to significantly benefit from using these libraries it would more or less have to be rewritten. There may still be value in some sort of integration layer, but that should be done as a separate project for those who already use one of these libraries.

That said, I still think that there should be an easier way to create object bindings. Probably the most straight-forward way to do this is to add a series of combinators:

```scala
def map[A1, R](
  a1: ObservableValue[A1],
  f: A1 => R
): ObjectBinding[R] =
  Bindings.createObjectBinding(() => f(a1.value), a1)

def map2[A1, A2, R](
  a1: ObservableValue[A1],
  a2: ObservableValue[A2],
  f: (A1, A2) => R
): ObjectBinding[R] =
  Bindings.createObjectBinding(() => f(a1.value, a2.value), a1, a2)

def map3[A1, A2, A3, R](
  a1: ObservableValue[A1],
  a2: ObservableValue[A2],
  a3: ObservableValue[A3],
  f: (A1, A2, A3) => R
): ObjectBinding[R] =
  Bindings.createObjectBinding(() => f(a1.value, a2.value, a3.value), a1, a2, a3)

// etc.
```

It should easily be possible to use [sbt-boilerplate](/sbt/sbt-boilerplate) to generate these. What do you think?

**Comment by jpsacha** [comment]
I see where your concern is. You are consider a case when arguments of the function that calculates the binding is an evaluation of a function that has the same as arguments as dependencies of the binding. In that case it would be nice to avoid repeating the arguments. 

Note that by type the function that you pass to `createObjectBinding` has no arguments. 
```scala
def createObjectBinding[T](func: () => T, dependencies: Observable*): : ObjectBinding[T]
```
A use case that I see the most frequently in project code is that `func` operated on private properties of an object, the `dependencies` are just used to trigger `func` execution. In other words, there is no call that looks like `f(a1.value, a2.value, a3.value)`

Your point is valid use case too. That will need some assumptions about the signature of `func`. What about something line this:
```scala
def createObjectBinding[T](func: (Seq[Observable]) => T, dependencies: Observable*): ObjectBinding[T]
```
You will pass a `func` that takes `dependencies` as argument, but you only specify `dependencies` once. Thoughts?

**Comment by mberndt123** [comment]
Hi @jpsacha,

Unfortunately I don't think a method like that would be particularly useful, because you can't really do anything with a `Seq[Observable]` – it's not possible to pull any values out because `Observable` doesn't have a value method, and if it did, it would have to return `Any`, which you can't do anything with (except with a cast).

**Comment by joncard1** [comment]
I'm sorry if it's bad manners to revive an old thread like this, but I came here to suggest this. At least on BufferProperty[T],

map[T, U]( T => U): BufferProperty[U]

that could be bound to another BufferProperty so that, when the first property changes, the second property gets the transformed content.

My use case is I am building a chart to display the data generated by a simulator. When a slider changes, I'd like the simulator to re-run and then chart the data. To setup, I have

```
val simulatorProperty: BufferProperty[MyDataset] = new BufferProperty()

heatFlowSlider.value.onChange((_, _, newValue) => {
    simulatorProperty().clear()
    simulatorProperty().addAll(generator.generatePoints(newValue.doubleValue()))
})
```
What I'd like is:

```
temperatureChart.data <== simulatorProperty.map[javafx.scene.chart.XYChart.Series[Number, Number]](dataset => {
  XYChart.Series[Number, Number](
    dataset.name,
    ObservableBuffer.from(
      dataset.temperatures.zipWithIndex.map(tuple =>
        XYChart.Data(tuple._2, tuple._1)
      )
    )
  )
})

```
What I think I have to have is:

```
val subscription = simulatorProperty().onChange((newValue, _) => {
  temperatureChart.data() =
      simulatorProperty()
         .map[javafx.scene.chart.XYChart.Series[Number, Number]](dataset => {
            XYChart.Series[Number, Number](
              dataset.name,
              ObservableBuffer.from(
                dataset.temperatures.zipWithIndex.map(tuple =>
                  XYChart.Data(tuple._2, tuple._1)
                )
              )
            )
        })
  })
}
```
(I have reasons to want the simulator results to go into an intermediate property rather than directly into the chart data)

Not really the end of the world. I appreciate I'm being a little neurotic, but is this an interesting idea? Would you be willing to review it if I added this to BufferProperty and submitted a pull request?

**Reply by jpsacha** [reply]
In JavaFX 19 `ObservableValue` has method 'map'. ScalaFX 19.0.0-R30 supports it, so you can write following to bind, for instance, an arbitrary property to a string property
```scala
textProperty <== modelProperty.map(v => v.toString)
```
In your example, `simulatorProperty` will now have that `map` method. Does it help you to address the issue?

**Comment by pljones** [comment]
Again, _similar_ to this topic but not directly related...  (Just posted on the ScalaFX GoogleGroup, which appears dead and has dead links - but is still accepting joining and posts...)

I've got a collection of values:
```scala
trait DataItem // this is where  val changed: BooleanProperty  is defined

abstract class Pad extends DataItem

abstract class PadSeq[TPad <: Pad] extends DataItem
  private val _pads: Array[TPad] = ...
  ...
  changed <== // something
```
So `Pad` and `PadSeq` both have the `BooleanProperty` called `changed`.  On `Pad`, this gets set to true on any update operation.

For `PadSeq`, I want `changed` bound to every `Pad.changed` property...

Now, I can say this:
```scala
changed <== _pads(0).changed || _pads(1).changed // ... through to end of array by hand
```
but I'd like to know if there's an array (or other) operator I can use to "reduce" the set of `Pad` `BooleanProperties` yet preserve the binding to each discrete value, so that `PadSeq.changed` updates correctly.

**Reply by pljones** [reply]
A more complete, simple example:
```scala
class _Bar extends BooleanProperty{
    def reset(): Unit = {
        value = false
    }

    protected def update(u: => Unit) = {
        u
        value = true
    }
}
class _Foo {
    private val _foo: ObservableBuffer[_Bar] = new ObservableBuffer[_Bar]()
    _foo.addAll(new _Bar(), new _Bar(), new _Bar()) // and more...
    val _baz: BooleanProperty = new BooleanProperty() { value = false }
    // _baz <== ??? for all of _foo
    _baz <== _foo(0) || _foo(1) || _foo(2) // ... "by hand" seems clumsy
}
```
Here, I've added an `ObservableBuffer` around the class with the boolean property (which now extends the BooleanProperty directly, rather than having it as a class variable).  It doesn't appear to add value, though.

**Reply by pljones** [reply]
And, to add, in many places, I'm using an idiom where one abstract class defines the collection of properties but there are several concrete classes with different size collections.  At the moment, this means each concrete class needs to do something like
```scala
abstract class MyClass(u: => Array[Foo]) {
  val _foo: Array[Foo] = u()
  val _other: DataItem = new DataItem() { ... } // using a concrete instance of DataItem
  changed <== changed || _other.changed
}
class MyClassA extends MyClass(() => new Array[Foo](1)) {
  changed <== changed || _foo(0).changed
}
class MyClassB extends MyClass(() => new Array[Foo](4)) {
  changed <== changed || _foo(0).changed || _foo(1).changed || _foo(2).changed || _foo(3).changed
}
```

**Reply by jpsacha** [reply]
@pljones Something like this may work:
```scala
  val e = BooleanProperty(false)
  val f = BooleanProperty(false)
  val g = BooleanProperty(false)
  val x = BooleanProperty(false)

  val bools = Seq(e, f, g)
  // Since we will be folding on OR, we will start with FALSE initial binding
  val dummyBinding = BooleanProperty(false) && BooleanProperty(false)
  x <== bools.fold(dummyBinding)(_ || _)
```
You need to initialize `fold` with a `BooleanBinding`, ten you fold `BooleanPropertie`s using `||`. Maybe there is some way to create empty binding or you could use initial elements of the sequence for that:
```scala
  x <== bools.drop(2).fold(bools(0) || bools(1))(_ || _)
```
first version may be cleaner anyway

**Reply by pljones** [reply]
I changed to
```
class PadSeq[TPad <: Pad] extends DataItem {
    private val _pads: Array[TPad] = ((0 to 27) map (f(_))).toArray
...
    _pads map(p => p.changed.onChange((_, _, newValue) => changed() |= newValue))
}
```
where `PadSeq` and `Pad` both extend `DataItem`, a trait with a `BooleanProperty` `changed`.  (i.e. a hard binding on each element of the array directly, rather than trying to aggregate into some "dynamic binding".)  Most of the _hard_ work I was having was the big arrays of DataItems I've got...  The lower-level data types have "update" and "apply" wrappers that set `changed` on their owner, rather than being `DataItems` themselves

But your example _does_ help explain some of the errors I was getting with my attempts!  Thanks.
